File: matlab/addnc.m
function		addnc(fname,X)

%		addnc(fname,X)
%		Add a variable to a NetCDF archive file. If the archive file does not exist,
%		it is created. The file is assumed to be in the current working directory 
%		unless a pathname is added to the beginning of fname.
%
%		Inputs:
%		fname is the name of the metadata file. If the name does not include a .nc
%		 suffix, this will be added automatically.
%		X is a sensor or metadata structure. Only these kind of variables can be saved
%		 in a NetCDF file because the supporting information in these structures is
%		 needed to describe the contents of the file. For non-archive and non-portable
%		 storage of variables, consider using the usual 'save' function in Matlab and Octabe.
%
%		Example:
%		 addnc('dog17_124a',A)
% 	    generates a file dog17_124a.nc and adds a variable A.
%
%     Valid: Matlab, Octave
%     markjohnson@st-andrews.ac.uk
%     last modified: 12 July 2017

if nargin<2,
	help savenc
end
	
if ~isstruct(X),
	fprintf('savenc can only save sensor or metadata structures\n') ;
	return
end
	
% append .nc suffix to file name if needed
if length(fname)<3 || ~all(fname(end+(-2:0))=='.nc'),
   fname(end+(1:3))='.nc';
end

% test if X is a metadata structure or a sensor structure
if ~isfield(X,'name')
	vname = [] ;
else
	vname = X.name ;
end

% check that the deployment ID of X matches the one in the file if the file
% already exists
depid = [] ;
if exist(fname,'file'),
	S = ncinfo(fname) ;
	k = find(strcmp({S.Attributes(:).Name},'depid')) ;
	if ~isempty(k),
		depid = S.Attributes(k).Value ;
		if strcmp(depid,X.depid)==0
			fprintf('File already associated with deployment id: %s. Choose a different file name.\n', depid);
			return ;
		end
	end
			
	% check if there is already a variable with this name in the file
	k = find(strcmp({S.Variables(:).Name},vname)) ;
	if ~isempty(k),
		fprintf('Variable %s already exists in file: choose a different name\n',vname) ;
		return
	end
end

% now ready to save the structure
if ~isempty(vname),		% X is a sensor structure
	if ~isfield(X,'data') || isempty(X.data),
		nccreate(fname,vname);
	else
		nccreate(fname,vname,'Dimensions',{'samples',size(X.data,1),'axis',size(X.data,2)});
		ncwrite(fname,vname,X.data);
	end

	ncwriteatt(fname,vname,'name',X.name);
	if isfield(X,'fs'),
		ncwriteatt(fname,vname,'sampling','regular');
		ncwriteatt(fname,vname,'sampling_rate',X.fs);
		ncwriteatt(fname,vname,'sampling_rate_unit','Hz');
	end
	
	if ~isfield(X,'meta'),
		fprintf('Warning: No metadata in variable %s\n',X.name);
	else
		F = fieldnames(X.meta) ;
		V = struct2cell(X.meta) ;
		for k=1:length(F),
			if ~ischar(V{k})
				fprintf('All metadata fields must be strings: leaving field %s blank\n',F{k}) ;
				ncwriteatt(fname,vname,['meta_' F{k}],'') ;
			else
				ncwriteatt(fname,vname,['meta_' F{k}],V{k}) ;
			end
		end
	end

	% save some default file attributes if none are present
	if isempty(depid),
		ncwriteatt(fname,'/','depid',X.depid);
		depid = X.depid ;
	end
	ncwriteatt(fname,'/','creation_date',datestr(now));
	return
end	

% Otherwise X is a metadata structure. Add it to the general attributes for the file
% Overwrite any field already present
if isempty(depid),
	nccreate(fname,'_empty');
end

F = fieldnames(X) ;
V = struct2cell(X) ;
for k=1:length(F),
	if ~ischar(V{k})
		fprintf('All metadata fields must be strings: leaving field %s blank\n',F{k}) ;
		ncwriteatt(fname,'/',F{k},'') ;
	else
		ncwriteatt(fname,'/',F{k},V{k}) ;
	end
end

ncwriteatt(fname,'/','creation_date',datestr(now));
return

File: R/forLater/add_nc.R
#'	Save an item to a NetCDF Add one tag sensor or metadata variable to a NetCDF archive file. file. 
#'		
#'	Add one tag sensor or metadata variable to a NetCDF archive file.If the archive file does not exist,
#'	it is created. The file is assumed to be in the current working directory 
#'	unless a pathname is added to the beginning of fname. 
#'	
#'	@param file The name of the netCDF file to which to save. If the name does not include a .nc suffix, this will be added automatically.
#'	@param X The sensor data or metadata list to be saved. a list of tag sensor and/or metadata lists. Alternatively, sensor and metadata lists may be input as multiple separate unnamed inputs. Only these kind of variables can be saved in a NetCDF file because the supporting information in these structures is needed to describe the contents of the file. For non-archive and non-portable storage of variables, consider using \code{\link{save}} or various functions to write data to text files. 
#'	@seealso \code{\link{save_nc}}, \code{\link{load_nc}}	
#'	@export
#'  @example
#'  \dontrun{
#'  add_nc('dog17_124a',A)
#'  # generates a file dog17_124a.nc (if it does not already exist) and adds a variable A.
#'  }

add_nc <-function(file,X){
  # input checking
  if(!is.list(X) | "animaltag" %in% class(X)){
    stop('save_nc can only save individual sensor or metadata structures.') 
  }

  # append .nc suffix to file name if needed
  if (!grepl('.nc', file)){
    file <- paste(file, '.nc', sep='')
  }
  
  # test if X is a metadata structure or a sensor structure
  if(length(X$depid) > 0 | length(X$sampling_rate)>0){
    
  }
  
  #get name of input data structure
  vname <- as.character(substitute(X))

  # check that the deployment ID of X matches the one in the file if the file
  # already exists
  prev_depid <- c() ;
  if(file.exists(file)){
    S <- ncdf4::nc_open(file)
    prev_depid <- ncdf4::ncatt_get( S , 0 )$depid
    if (length(prev_depid) != 0){
      if (!identical(prev_depid,X$depid)){
        e_msg <- paste('Chosen file name is already associated with deployment id: ',
                       prev_depid, '. Choose a different file name.\n',
                       sep='')
        stop(e_msg)
      }
    }
    
    prev_vars <- names(S$var)
    if(vname %in% prev_vars){
      e_msg <- paste('Variable ', vname,
                     ' already exists in file. Choose a different name.\n',
                     sep='')
      stop(e_msg)
    }
  }# end of "if file already exists" checks

  # now ready to save the data or metadata
  if ('data' %in% names(X)){ # X is a sensor structure
    if (length(X$data) == 0){
      # if X is empty...
      ncv <- ncdf4::ncvar_def(name=vname, 
                              units='',#X$meta_unit,
                              dim=list(),
                              missval=NA)
      ncdf4::nc_create(file,ncv)
    }
    else{ #if there is some data
      dims <- list(ncdf4::ncdim_def(name='samples', 
                                    units='number', 
                                    vals=c(1:dim(X$data)[1])),
                   ncdf4::ncdim_def(name='axis', 
                                    units='number',
                                    vals=c(1:dim(X$data)[2])))
      
      ncv <- ncdf4::ncvar_def(name=vname, 
                              units=X$meta_unit,
                              dim=dims,
                              missval=NA,
                              longname=X$meta_full_name)
      if (!exists(file)){
        # if the file doesn't exist create it with this var
        nc_conn <- ncdf4::nc_create(file,ncv)
      }else{
        # if file already exists add variable to it
        nc_conn <- ncdf4::nc_open(file, write=TRUE)
        ncdf4::ncvar_put(nc_conn, ncv, 
                         vals=as.vector(X$data),
                         count=dim(X$data))
      }
    }# end of writing sensor data
    
    # add metadata (from sensor data structure)
    i_meta <- which(names(X)!='data')
    for (m in i_meta){
      ncdf4::ncatt_put(nc_conn, varid=ncv, 
                       attname=names(X)[m],
                       attval=X[[m]])  
    }

    if (length(i_meta==0)){
      w_msg <- paste('No metadata in variable ',
                     vname, '.\n')
      warning(w_msg)
    }
        
    if (length(prev_depid) == 0){
      #if this is a new file make sure the depid is specified
      # here the default is to use the file name string as depid
      ncdf4::ncatt_put(nc_conn, varid=ncv, 
                       attname='depid',
                       attval=sub('.nc', '', basename(file)))
    }
  }	#end of "if it's a sensor data structure	
 
  # Otherwise X is a metadata structure. Add it to the general attributes for the file
  # Overwrite any field already present
  if(length(depid) == 0){
    ncdf4::nc_create(fname,'_empty');
  }
  #need to use with new testsets above; they have depid in the sensor data.
  # check for other changes too?
  
  
  F = fieldnames(X) 
  V = struct2cell(X) 
  for (k in 1:length(F)){
    if(!ischar(V[[k]])){
      fprintf('All metadata fields must be strings: leaving field #s blank\n',F[[k]]) ;
      ncwriteatt(fname,'/',F[[k]],'') 
    }
    else{
      ncwriteatt(fname,'/',F[[k]],V[[k]]) ;
    }
  }

  # add creation date (whether it was sensor or meta)
  ncdf4::ncatt_put(nc_conn, ncv,
                   attname='creation_date',
                   attval=Sys.time());
  
  ncdf4::nc_close()
}
